File: leetcode/134-gas-station.js
// https://leetcode.com/problems/gas-station/ (top-interview-questions)

/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
  const diff = [];
  let totalDiff = 0;

  for (let i = 0; i < gas.length; i += 1) {
    diff.push(gas[i] - cost[i]);
    totalDiff += (gas[i] - cost[i]);
  }

  // gas total sum < cost total sum => so we have not enough cas to make a tour
  if (totalDiff < 0) return -1;

  let tank = 0;
  let start = 0;
  for (let i = 0; i < gas.length; i += 1) {
    tank += diff[i];
    if (tank < 0) {
      start = i+1;
      tank = 0;
    }
  }

  return start;
};

File: leetcode/134-gas-station.py
# https://leetcode.com/problems/gas-station/ (top-interview-questions)

from typing import List


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        # if the total existing quantity of gas in all the stations
        # is less then the road cost of gas then it's impossibale to
        # travel around the circuit
        if sum(gas) < sum(cost):
            return -1
        
        tank = 0 # the quantity of gas in the tank
        start = 0 # the starting position
        for i in range(len(gas)):
            tank += gas[i] - cost[i]
            if tank < 0:
                tank = 0
                start = i + 1
        
        return start
